sql/moz-fx-data-shared-prod/org_mozilla_ios_firefox/unified_metrics/query.py

```diff
@@ -0,0 +1,216 @@
+"""Generate a query for unifying legacy metrics and glean metrics.
+
+This approach is required because the order of columns is significant in order
+to be able to union two tables, even if the set of columns are the same."""
+
+import requests
+import json
+from bigquery_etl.format_sql.formatter import reformat
+from google.cloud import bigquery, exceptions
+import io
+
+
+def get_columns(schema):
+    """Return a list of columns corresponding to the schema.
+
+    Modified from https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/c2dae92a7ed73e4774a897bc4d2a6a8608875339/mozilla_pipeline_schemas/utils.py#L30-L43
+    """
+
+    def traverse(prefix, columns):
+        res = []
+        for node in columns:
+            name = node["name"]
+            dtype = node["type"]
+            # we consider a repeated field a leaf node, and sorted for our purposes
+            if dtype == "RECORD" and node["mode"] != "REPEATED":
+                res += traverse(f"{prefix}.{name}", node["fields"])
+            else:
+                res += [f"{prefix}.{name} {dtype}"]
+        return res
+
+    res = traverse("root", schema)
+    return sorted(res)
+
+
+def generate_query(columns, table):
+    """Generate a SQL query given column names.
+
+    We construct a query that selects columns into nested structs. Naive selection
+    of all the columns will strip the namespace from the columns.
+    """
+
+    # Build a string that contains the selected columns. We take the set of
+    # columns and split them up by namespace. Each namespace is put inside of a
+    # STRUCT call. For example, foo.a and foo.b will be translated into a
+    # `STRUCT(foo.a, foo.b) as foo` nested column.
+    acc = ""
+
+    # Maintain the last element in the columns to determine when a transition
+    # must be made.
+    prev = []
+
+    # Iterate over the sorted set of columns. This ensures that columns are
+    # grouped together correctly. Every time the column goes into a namespace,
+    # we push an opening struct statement onto the string. Every time we
+    # complete nested struct, we close out the string by aliasing the struct to
+    # the namespace.
+    for col in sorted(columns):
+        split = col.split(".")
+        # check if we go deeper
+        if len(split) > 1 and len(split) > len(prev):
+            # the number of times to start nesting
+            if len(prev) == 0:
+                k = len(split) - 1
+            else:
+                k = len(split) - len(prev)
+            acc += "struct(" * k
+        # sometimes we have two structs that are the same depth e.g. metrics
+        if len(split) > 1 and len(split) == len(prev) and split[-2] != prev[-2]:
```

Comment:
`a.b.c.d`, `a.e.f.g` => this section assumes they share the ancestor `e`. It should be more like: `while split[-i] != prev[-i]`, and then finish off the previous struct.

This is a real quagmire. Even with that adjustment, strings like `a.b.c.d.e` and `a.f.c.g.h` will fail, because the algo will see `c`, think it's done, but miss the parent `b`/`f` difference.

I see two options here:
1. Fix this algorithm to build up from the bottom (`a`/`a`, then `b`/`e` differ)
2. Clearly identify these flaws, state this is not for general purpose use yet, and do some sort of processing at the beginning to ensure the columns are part of an accepted set (i.e. glean-shaped data only, where we know max depth and comparisons).

Reply:
This is certainly an issue, it does need to consider the base case for two nested structs like metadata and metrics.

Reply:
Does this new method satisfy the test of `a.b.c.d` and `a.f.c.g`?

Reply:
No :(

I wish there were a way to just generate all the possible cases of depth 4 with their solutions to test the python code.

Reply:
This new method satisfies the test case that you brought up.

Reply:
Hah, then we would use that function for the code itself!

Why does this require the prev and cur be the same length? How about `a.b.c.d` and `a.b.e`? Sorry for all these edge cases :/

Reply:
I want to say this will be okay, since we'll pop out of that nested level, and then handle it assuming that the two columns are the same level. I'll add the test case though.